WordPress/src/main/java/org/wordpress/android/ui/reader/repository/usecases/PostLikeUseCase.kt

```diff
@@ -29,86 +24,101 @@ import kotlin.coroutines.Continuation
 
-            if (isAskingToLike) {
-                analyticsUtilsWrapper.trackWithReaderPostDetails(READER_ARTICLE_LIKED, post)
-                // Consider a like to be enough to push a page view - solves a long-standing question
-                // from folks who ask 'why do I have more likes than page views?'.
-                readerPostActionsWrapper.bumpPageViewForPost(post)
-            } else {
-                analyticsUtilsWrapper.trackWithReaderPostDetails(READER_ARTICLE_UNLIKED, post)
-            }
+        handleLocalDb(post, request)
+    }
 
-            return@withContext suspendCancellableCoroutine<ReaderRepositoryCommunication> { cont ->
-                continuations[request] = cont
-                readerPostActionsWrapper.performLikeAction(post, isAskingToLike, wpComUserId)
-            }
+    private suspend fun FlowCollector<PostLikeState>.handleLocalDb(
+        post: ReaderPost,
+        request: PostLikeRequest
+    ) {
+        val response = readerPostActionsWrapper.performLikeActionLocal(
+                post,
+                request.isAskingToLike,
+                request.wpComUserId
+        )
+        if (response) {
+            emit(PostLikeState.PostLikedInLocalDb(request))
+            val postLikeState = performLikeAndWaitForResult(post, request)
+            emit(postLikeState)
+        } else {
+            emit(Unchanged(request)) // this can be considered a success
         }
     }
 
-    @Subscribe(threadMode = BACKGROUND)
-    @SuppressWarnings("unused")
-    fun onPerformPostLikeEnded(event: ReaderRepositoryEvent) {
-        if (event is PostLikeEnded) {
-            val request = PostLikeRequest(
-                    event.postId,
-                    event.blogId,
-                    event.isAskingToLike,
-                    event.wpComUserId
-            )
-
-            val comm: ReaderRepositoryCommunication = when (event) {
-                is PostLikeUnChanged -> SuccessWithData(event)
-                is PostLikeSuccess -> SuccessWithData(event)
-                is PostLikeFailure -> RemoteRequestFailure
+    private suspend fun performLikeAndWaitForResult(
+        post: ReaderPost,
+        request: PostLikeRequest
+    ): PostLikeState {
+        val actionListener = ActionListener { succeeded ->
+            val postLikeState = if (succeeded) {
+                Success(request)
+            } else {
+                RequestFailed(request)
             }
-
-            continuations[request]?.resume(comm)
+            continuations[request]?.resume(postLikeState)
             continuations[request] = null
         }
+
+        return suspendCancellableCoroutine { cont ->
+            continuations[request] = cont
+            readerPostActionsWrapper.performLikeActionRemote(
+                    post,
+                    request.isAskingToLike,
+                    request.wpComUserId,
+                    actionListener
+            )
+        }
     }
 
     fun stop() {
```

Comment:
It seems this method is not used. Moreover, I think it's a bit misleading - calling "clear" won't stop anything, it'll just "deadlock" a few coroutines as no-one ever resumes them. Wdyt?

Reply:
> It seems this method is not used. Moreover, I think it's a bit misleading - calling "clear" won't stop anything, it'll just "deadlock" a few coroutines as no-one ever resumes them. Wdyt?

Makes sense. Removed in commit ccfcc11